core/src/main/java/io/questdb/cairo/CairoEngine.java

```diff
@@ -55,92 +56,108 @@
-    private final long tableIndexMemSize;
+    private final RingQueue<TableWriterTask> tableWriterCmdQueue;
+    private final MCSequence tableWriterCmdSubSeq;
+    private final long tableIdMemSize;
+    private long tableIdFd = -1;
+    private long tableIdMem = 0;
 
     public CairoEngine(CairoConfiguration configuration) {
         this.configuration = configuration;
         this.messageBus = new MessageBusImpl(configuration);
         this.writerPool = new WriterPool(configuration, messageBus);
         this.readerPool = new ReaderPool(configuration);
         this.writerMaintenanceJob = new WriterMaintenanceJob(configuration);
-        this.telemetryQueue = new RingQueue<>(TelemetryTask::new, configuration.getTelemetryConfiguration().getQueueCapacity());
-        this.telemetryPubSeq = new MPSequence(telemetryQueue.getCapacity());
-        this.telemetrySubSeq = new SCSequence();
-        telemetryPubSeq.then(telemetrySubSeq).then(telemetryPubSeq);
-
-        final FilesFacade ff = configuration.getFilesFacade();
-        try (Path path = new Path().of(configuration.getRoot()).concat("_tab_index.d").$()) {
-            this.tableIndexMemSize = Files.PAGE_SIZE;
-            tableIndexFd = TableUtils.openFileRWOrFail(ff, path);
-            final long fileSize = ff.length(tableIndexFd);
-            if (fileSize < Long.BYTES) {
-                if (!ff.allocate(tableIndexFd, Files.PAGE_SIZE)) {
-                    ff.close(tableIndexFd);
-                    throw CairoException.instance(ff.errno()).put("Could not allocate [file=").put(path).put(", actual=").put(fileSize).put(", desired=").put(this.tableIndexMemSize).put(']');
-                }
-            }
-
-            this.tableIndexMem = ff.mmap(tableIndexFd, tableIndexMemSize, 0, Files.MAP_RW);
-            if (tableIndexMem == -1) {
-                ff.close(tableIndexFd);
-                throw CairoException.instance(ff.errno()).put("Could not mmap [file=").put(path).put(']');
-            }
-            try {
-                new EngineMigration(this, configuration).migrateEngineTo(ColumnType.VERSION);
-            } catch (CairoException e) {
-                close();
-                throw e;
-            }
+        if (configuration.getTelemetryConfiguration().getEnabled()) {
+            this.telemetryQueue = new RingQueue<>(TelemetryTask::new, configuration.getTelemetryConfiguration().getQueueCapacity());
+            this.telemetryPubSeq = new MPSequence(telemetryQueue.getCycle());
+            this.telemetrySubSeq = new SCSequence();
+            telemetryPubSeq.then(telemetrySubSeq).then(telemetryPubSeq);
+        } else {
+            this.telemetryQueue = null;
+            this.telemetryPubSeq = null;
+            this.telemetrySubSeq = null;
+        }
+        this.tableIdMemSize = Files.PAGE_SIZE;
+        // subscribe to table writer commands to provide cold command handling
+        this.tableWriterCmdQueue = messageBus.getTableWriterCommandQueue();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code assigns a resource that might leak to a class field. This can cause a resource leak that slows down or crashes your system. For example, when EngineMigration.migrateEngineTo throws an exception, the *CairoEngine* object isn’t constructed so you can’t close this resource afterward.

**Fix**
Ensure that if an exception is thrown, the *tableWriterCmdQueue* resource is closed before the constructor is exited.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).